File: 1056 Degrees of Separation/solution.java
import java.io.*;
import java.util.*;

public class Main {
	static int INF = 50;
    static int[][] dist = new int[50][50];
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
		int dataSets = 0, P;
		StringBuilder output = new StringBuilder();
		while(st.nextToken() == StreamTokenizer.TT_NUMBER && (P = (int)st.nval) != 0) {
			for(int i = 0; i < P; ++i) {
				for(int j = 0; j < P; ++j)
					dist[i][j] = i == j ? 0 : INF;
			}
			st.nextToken();
		    int R = (int)st.nval;
		    HashMap<String, Integer> peopleID = new HashMap<>();
		    int id = 0;
		    while(R-- > 0) {
		    	st.nextToken();
		    	String name1 = st.sval;
		    	st.nextToken();
		    	String name2 = st.sval;
		    	if(!peopleID.containsKey(name1))
		    		peopleID.put(name1, id++);
		    	if(!peopleID.containsKey(name2))
		    		peopleID.put(name2, id++);
		    	dist[peopleID.get(name1)][peopleID.get(name2)] = dist[peopleID.get(name2)][peopleID.get(name1)] = 1;
		    }
		    floydWarshall(P);
		    int max = 0;
		    boolean disconnected = false;
		    for(int i = 0; i < P && !disconnected; ++i) {
		    	for(int j = 0; j < P && !disconnected; ++j)
		    		if(dist[i][j] == INF)
		    			disconnected = true;
		    		else
		    			max = Math.max(max, dist[i][j]);
		    }
		    output.append("Network ").append(++dataSets).append(": ").append(disconnected ? "DISCONNECTED" : max).append("\n\n");
		}
		System.out.print(output);
	}
	
	static void floydWarshall(int V) {
		for(int k = 0; k < V; ++k) {
			for(int i = 0; i < V; ++i) {
				for(int j = 0; j < V; ++j)
					dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
			}
		}
	}
}

File: 1056 Degrees of Separation/solution.cpp
#include <bits/stdc++.h>
using namespace std;

const int INF = 50;
vector<vector<int>> dist(50, vector<int>(50));

void floydWarshall(int V) {
    for(int k = 0; k < V; ++k) {
        for(int i = 0; i < V; ++i) {
            for(int j = 0; j < V; ++j)
                dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    int dataSets = 0, P, R;
    while(cin >> P >> R && P != 0) {
        for(int i = 0; i < P; ++i) {
            for(int j = 0; j < P; ++j)
                dist[i][j] = i == j ? 0 : INF;
        }
        unordered_map<string, int> peopleNum;
        string name1, name2;
        while(R--) {
            cin >> name1 >> name2;
            if(!peopleNum.count(name1))
                peopleNum[name1] = peopleNum.size();
            if(!peopleNum.count(name2))
                peopleNum[name2] = peopleNum.size();
            dist[peopleNum[name1]][peopleNum[name2]] = dist[peopleNum[name2]][peopleNum[name1]] = 1;
        }
        floydWarshall(P);
        int ans = 1;
        for(int i = 0; i < P; ++i) {
            for(int j = 0; j < P; ++j)
                ans = max(ans, dist[i][j]);
        }
        if(ans == INF)
            cout << "Network " << ++dataSets << ": DISCONNECTED\n\n";
        else
            cout << "Network " << ++dataSets << ": " << ans << "\n\n";
    }
    return 0;
}
